### DISCUSSION TITLE
Issue JWT or Bearer token for specific user programmatically with management API

### DISCUSSION THREAD

**Discussion by sm3142** [zitadel/zitadel#8406] [General]
Hi all,

I have a somewhat unusual requirement for authenticating clients _and_ an idea how to solve it, but I'm not sure it can be done with ZITADEL (at least presently).

The premise is that the client is an IoT device with a TPM module, which is seeded with an X.509 certificate/private key at the factory. Furthermore, it's not so easy to push additional client-specific information to the device in a secure way, e.g. things like PATs, JWT public/private keys or good old passwords. In the end, the IOT devices need to access API services secured with ZITADEL in the usual manner, such that the service can establish the identity and authorization of the requesting device.

What I have in mind is to "exchange" the credential the device already havs (x.509 cert/private key) with a ZITADEL-issued session token for subsequent API access. To that end, a ZITADEL service (machine) user is created for each device and the relation of device ID (respectively X.509 subject name) and ZITADEL user ID is stored in a database lookup table.

The credential exchange should then roughly work like this:

1. At bootup, the IoT device calls a special API endpoint* requiring mutual TLS (mTLS), thus coercing the client into presenting a TLS certificate (signed with he private key in the TPM) of its own.
2. The endpoint (or the API gateway in front of it) verifies the validity of the presented client certificate and accepts / rejects the request accordingly.
3. If accepted, the endpoint looks up the corresponding ZITADEL user ID for the given certificate subject name.
4. If found, the endpoint then **requests a session token for the given user ID through the ZITADEL management API and returns it to the client device**.
5. For the remainder of its uptime, the device subsequently uses the session token to authorize any other API calls.

What I do not know (and couldn't figure out from the documentation) is whether 4. is even possible with ZITADEL. The rest is essentially just busywork - not exactly trivial but eminently doable. I'm aware that there are various other ways to achieve something similar, such as exchanging for a PAT or another form of client credential. However, the reason I'm interested in the method sketched out above is that it doesn't involve storing fixed credentials in clear text or sending them over the wire at any time.

Any pointers appreciated (even if it's a flat out can't-be-done-with-ZITADEL)!
-stefan

*) just to be exceedingly clear here, this is not expected to be provided by ZITADEL, but a custom service of our own making.

**Comment by fforootd** [comment]
Hey, thanks for this question.

I think it is feasible to pull this of with a service in front of the session api (or maybe token exchange could also work)

So basically your component does mTLS and then sends the user to zitadel session api to get a token.

Main drawback with this certainly is that the user is not "authenticated" in zitadel this way 😁

**Reply by sm3142** [reply]
@fforootd, thanks for your response! I assume you mean this guy here: https://zitadel.com/docs/apis/resources/session_service_v2/session-service-create-session ? I was looking at this previously but it's not clear to me how this would work without having the user credentials - i.e. is it possible to create a session without providing anything other than the `user` in the `checks` section of the body? And what grants would the caller need to have? The documentation is not very forthcoming unfortunately.

> (or maybe token exchange could also work)

This seems more promising to me, especially [Impersonation by user ID example](https://zitadel.com/docs/guides/integrate/token-exchange#impersonation-by-user-id-example). Unfortunately I can't get any of the examples working on that page working (see also my message on the [ZITADEL Discord]( https://discord.com/channels/927474939156643850/1272587161060900885/1272587161060900885) 😥

> Main drawback with this certainly is that the user is not "authenticated" in zitadel this way 😁

Yeah, I know, but it's the next best thing. As far as the backend is concerned, we still get all the trappings of a ZITADEL user, i.e. role-based access control, life-cycle management, etc. as with any other ZITADEL user. Just without having to copy awkward information (PATs, private keys, passwords, etc.) to the devices first (you may be surprised by how many IoT devices just use a shared key and the backend just has to trust that the device's identity is whatever the device claims it is). For sure, it _is_ a compromise, but IMO it's a near-optimal scenario.

**Reply by fforootd** [reply]
On the session api you should be able to just create a session with the userid or username as check.

Let me look into the token exchange thread in discord, that somewho slipped me 😁

**Reply by sm3142** [reply]
[fforootd](https://github.com/fforootd) thanks for your reply.

In the end I had more success using token exchange, specifically [Impersonation by user ID example](https://zitadel.com/docs/guides/integrate/token-exchange#impersonation-by-user-id-example). I have this working in a very rough PoC, but I yet have to translate the individually hand-crafted curl commands into actual server-side code.

Tried it with session creation as well, but somehow I didn't manage to translate the session token into an access token - I'm sure that's because I don't fully grasp the entire flow. In that sense token exchange is a more straightforward route for me.

**Comment by muhlemmer** [comment]
Hi @sm3142, allow me to reopen the discussion and continue it here. IMO github offers a better way than discord for others to search on similar issues in the future.

## mTLS

We currently do not support mTLS and it is not on our roadmap. You can open a feature request for it, but given the fact we didn't hear much demand for it, it might take a long while before we will implement it. Unless you company / product is big enough for an enterprise contract ;).

But, keep reading for other options.

## Which token?

First lets establish which type of token you _really_ want.

### Session token

Firstly lets talk about "session token". If you mean the token returned by the [Session API](https://zitadel.com/docs/apis/resources/session_service_v2/session-service) there is no way to exchange any credential, like a token, to a session token. This token type is only meant for session state management and is supposed to be exchanged to a OIDC access and ID token. However a session token _can_ be used for zitadel APIs. That possibility is mostly provided to allow user updates during the login flow. I don't think you want a session token, right?

### Access token

Instead, I think you want an access token, which can be used to send to your and zitadel's APIs. Your APIs can use [Introspection](https://zitadel.com/docs/apis/openidoauth/endpoints#introspection_endpoint) against zitadel to verify validity of that token.

## How to

Currently zitadel defines 2 main types of [users](https://zitadel.com/docs/concepts/structure/users):

- human users
- service users

As I read in your description a service user is all you need, so let's focus on that first. Where an optional human authentication can be added as an advanced case.

### Service user

To authenticate a service user, I do not recommend using token exchange (at the moment). We do plan to support [externally created token](https://github.com/zitadel/zitadel/issues/7211), perhaps self-signed, but we didn't go as far with that definition. That's because self-signed tokens can already be used in the [JWT profile grant](https://zitadel.com/docs/guides/integrate/service-users/private-key-jwt).

As you mentioned the device can sign certificates using a pre-loaded x.509 key with the TPM. Perhaps it can sign a JWT as well? If that's the case the following can be done:

At the factory:

1. create a service user: https://zitadel.com/docs/apis/resources/mgmt/management-service-add-machine-user
2. For each pre-loaded private key, create a public key. 
3. Set the public key(s) on the machine user: https://zitadel.com/docs/apis/resources/mgmt/management-service-add-machine-key. The public key data must be a PEM-encoded RSA public key in PKIX, ASN.1 DER form.

In the field during runtime:

1. Device creates a self-signed JWT, see [this page on the claims ](https://zitadel.com/docs/apis/openidoauth/grant-types#using-jwts-as-authorization-grants) that must be in the JWT. Depending which language the firmware uses, there might be a library that can help: https://jwt.io/libraries. 
2. Send a request to the token endpoint with the [JWT profile grant type](https://zitadel.com/docs/apis/openidoauth/endpoints#jwt-profile-grant). By setting the `urn:zitadel:iam:org:project:id:zitadel:aud` scope the token will be valid for ZITADEL APIs as well. You can set other scopes as needed to extend the audience to other projects, apps and APIs defined in zitadel.

Access tokens expire, so you do need to repeat the above to get a new token before that happens.

### Human user

If needed, human users can be authenticated two ways:

1. With the access token obtained for the service user above, impersonation can now be done as per the documentation you already found
2. If the device has some kind of display that can at least print URL to your instance, [Device authorization](https://zitadel.com/docs/guides/integrate/login/oidc/device-authorization) can be an option for a direct user login. In that case you could still use a self-signed JWT for c[lient_assertion authentication](https://zitadel.com/docs/apis/openidoauth/authn-methods#jwt-with-private-key) if you wish, or completely omit this step.

Hope this helps.

**Reply by sm3142** [reply]
Hello [muhlemmer](https://github.com/muhlemmer),

Thanks for the comprehensive answer! And sorry for the late reply - the notification got lost in my daily Github message deluge 🫣)

As for your suggested solution, I _think_ I understand it in broad strokes and it seems to make sense for my use case. Can't say I have experience with self-signed JWTs, but I to try to muddle through with help from the usual  places. It does involve a bit of supply chain gymnastics that I'd rather hoped to avoid (i.e. scraping the public keys & registering them with ZITADEL etc.), but nothing that can't be fixed with a script or two. On balance, this approach requires the least amount of moving parts or semi-exotic technology though. 

When I have any, I'll try to write up my results for posterity. In the meantime, thanks for the pro tip!

PS: To confirm some of your assumptions, for the benefit of other readers:
> Instead, I think you want an access token, which can be used to send to your and zitadel's APIs.

Yes, what's ultimately needed is an access token to access a custom API (i.e. a ZITADEL client application)

> As I read in your description a service user is all you need

Correct, our particular use case is strictly M2M - squishy humans need not apply 🤖 (yes, they are _actual_ robots :) 

> Perhaps it can sign a JWT as well?

I see no obstacles to this - the platform is running Linux, so all the necessary tooling should be readily available.

**Reply by muhlemmer** [reply]
You're welcome!

Offtopic:

> squishy humans need not apply 🤖 (yes, they are actual robots 

I saw your website. You guys do awesome stuff. I grew up in Rotterdam Botlek and worked in the offshore industry in a previous career. I can really understand the problems you solve with your nifty drones!

**Reply by sm3142** [reply]
After a few hours of trial & error I got the second part - the [JWT Profile Grant](https://datatracker.ietf.org/doc/html/rfc7523) flow - working. Since the documentation doesn't provide a full how-to, here's a small one for posterity.

1. Create a service user in the ZITADEL UI
2. Create key and download the JSON key file
3. Extrect the private key from the JSON key file:

    ```
    jq -r .key <JSON key file> privatekey.pem
    ```
4. Generate a signed profile JWT in your weapon of choice. Here's an example in Go:

    ```
    package main
    
    import (
	    "crypto/x509"
	    "encoding/pem"
	    "fmt"
	    "log"
	    "os"
	    "time"
    
	    "github.com/golang-jwt/jwt/v5"
    )
    
    const (
	    keyFile = <path to private key PEM file>
	    keyID   = <service user key ID>
	    userID  = "<service user id>"
	    domain  = "<ZITADEL instance domain in https://host.domain.tld form"
    )
    
    func main() {
	    pemString, err := os.ReadFile(keyFile)
	    if err != nil {
		    log.Fatalf("failed to read private key: %v", err)
	    }
	    block, _ := pem.Decode(pemString)
    
	    key, err := x509.ParsePKCS1PrivateKey(block.Bytes)
	    if err != nil {
		    log.Fatalf("failed to parse private key: %v", err)
	    }
    
	    token := jwt.NewWithClaims(jwt.SigningMethodRS256, jwt.MapClaims{
		    "aud":   domain,
		    "exp":   time.Now().Add(time.Second * 60).Unix(),
		    "iat":   time.Now().Unix(),
		    "iss":   userID,
		    "sub":   userID,
	    })
    
	    token.Header["kid"] = keyID
    
	    tokenString, err := token.SignedString(key)
	    if err != nil {
		    log.Fatalf("failed to sign token: %v", err)
	    }
    
	    fmt.Println(tokenString)
    }
    ```
4. The output of the program above can then be directly used in the POST request to the token endpoint:

    ```
    CUSTOM_DOMAIN=<ZITADEL custom domain in host.domain.tld form"
    curl --request POST \                                                                                                                                                      
      --url https://$CUSTOM_DOMAIN/oauth/v2/token \
      --header 'Content-Type: application/x-www-form-urlencoded' \
      --data grant_type=urn:ietf:params:oauth:grant-type:jwt-bearer \
      --data 'scope=openid profile email offline_access urn:zitadel:iam:org:project:id:<project_id>:aud' \
      --data assertion=`go run .`
    ```

    Scopes have to be added to the request as per usual, including any project-specific ones. Note that `offline_access` seems to be  spurious - it works without and no refresh token is provided even if you do add it. Maybe someone from ZITADEL can clarify this latter point?

However, this doesn't cover having the JWT signed by the TPM rather than a key on file. So far I've not been able to create a user key with a known public key, so there is no point testing the TPM signing yet. There already exists a TPM extension for `go-jwt` though: https://github.com/salrashid123/golang-jwt-tpm

I'll detail my troubles with uploading a public key in a separate post.

**Reply by sm3142** [reply]
As alluded in my earlier response, I couldn't get the key creation with a known public key as per the [documentation](https://zitadel.com/docs/apis/resources/mgmt/management-service-add-machine-key) working for the life of me. However, I have no real interest in doing this over the REST API anyway, so another small PoC in Go later, I have this working too:

```
package main

import (
	"context"
	"log"
	"os"
	"time"

	"github.com/zitadel/oidc/pkg/oidc"
	"github.com/zitadel/zitadel-go/v3/pkg/client/management"
	"github.com/zitadel/zitadel-go/v3/pkg/client/middleware"
	"github.com/zitadel/zitadel-go/v3/pkg/client/zitadel"
	"github.com/zitadel/zitadel-go/v3/pkg/client/zitadel/authn"
	mgmt_pb "github.com/zitadel/zitadel-go/v3/pkg/client/zitadel/management"
	"google.golang.org/protobuf/types/known/timestamppb"
)

const (
	issuer    = "<ZITADEL issuer in https://host.domain.tld form>"
	api       = "ZITADEL API string in "host.domain.tld:port" form>"
	publicKey = "<path to private key PEM file>"
	tokenFile = "<JSON key file of API user>"
	userID    = "<ID of user to add key too"
)

func main() {
	publicKey, err := os.ReadFile(publicKey)

	if err != nil {
		log.Fatal("failed to read public key", err)
	}

	ctx := context.Background()
	mc, err := management.NewClient(
		ctx,
		issuer,
		api,
		[]string{oidc.ScopeOpenID, zitadel.ScopeZitadelAPI()},
		zitadel.WithJWTProfileTokenSource(middleware.JWTProfileFromPath(ctx, tokenFile)),
	)

	if err != nil {
		log.Fatal("failed to create management client", err)
	}

	resp, err := mc.AddMachineKey(ctx, &mgmt_pb.AddMachineKeyRequest{
		UserId:         userID,
		Type:           authn.KeyType_KEY_TYPE_JSON,
		ExpirationDate: timestamppb.New(time.Now().Add(time.Hour * 24 * 365 * 20)),
		PublicKey:      publicKey,
	})

	if err != nil {
		log.Fatal("failed to add service user key", err)
	}

	log.Printf("service user key added: %v", resp)
}
```
The required key pair can be created with:
```
# generate key pair
openssl genrsa -out privatekey.pem 2048

# extract public key
openssl rsa -in keypair.pem -pubout -out publickey.pem
```
and the keys can then be used as-is to add the service user key and sign the JWT. The only thing that changes for the JWT signing code is that the private key format is slightly different than the one generated by ZITADEL. The following line:
```
key, err := x509.ParsePKCS1PrivateKey(block.Bytes)
```
then becomes:
```
key, err := x509.ParsePKCS8PrivateKey(block.Bytes)
```

PS: I'm sure the REST call can be made to work as well - probably something wrong with the bearer token I've used.

**Comment by sm3142** [comment]
[muhlemmer](https://github.com/muhlemmer), I suggest we wrap this up now. The TPM signing part has nothing to do with ZITADEL proper and is thus out of scope. Thanks again for pointing me in the right direction!

**Reply by muhlemmer** [reply]
You're welcome! I'll also try to find some time to go over your feedback on the documentation and see where we can improve.
